File: src/main/java/org/example/PathFinder.java
package org.example;

import java.util.HashSet;

public class PathFinder {

    public static boolean solution(char[][] A) {

        int numRows = A.length;
        int numColumns = A[0].length;

        int starti = -1;
        int startj = -1;

        // initializing a
        boolean[][] B = new boolean[numRows][numColumns];

        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < numColumns; j++) {
                if (A[i][j] != '.') {
                    B[i][j] = true;
                }

                if (A[i][j] == 'A') {
                    B[i][j] = false;
                    starti = i;
                    startj = j;
                }

                if (A[i][j] == 'v') {
                    int k = i+1;
                    while (k < numRows && A[k][j] == '.') {
                        B[k][j] = true;
                        k++;
                    }
                }

                if (A[i][j] == '^') {
                    int k = i-1;
                    while (k >= 0 && A[k][j] == '.') {
                        B[k][j] = true;
                        k--;
                    }
                }

                if (A[i][j] == '>') {
                    int k = j+1;
                    while (k < numColumns && A[i][k] == '.') {
                        B[i][k] = true;
                        k++;
                    }
                }

                if (A[i][j] == '<') {
                    int k = j-1;
                    while (k >= 0 && A[i][k] == '.') {
                        B[i][k] = true;
                        k--;
                    }
                }
            }
        }

        return f(starti, startj, numRows, numColumns, new HashSet<String>(),B);
    }

    public static boolean f(int i, int j, int numRows, int numColumns, HashSet<String> visited, boolean[][] B) {
        if (i < 0) return false;
        if (j < 0) return false;
        if (i >= numRows) return false;
        if (j >= numColumns) return false;
//        if (visited.contains([i, j])){
//            return false;
//        }
        if (B[i][j]) return false;
        if (j == numColumns-1) return true; //watch out

        visited.add(i + "," + j);
        return f(i, j+1,numRows,numColumns, visited,B) ||
                f(i, j-1, numRows,numColumns, visited,B) ||
                f(i+1, j, numRows,numColumns, visited,B) ||
                f(i-1, j, numRows,numColumns, visited,B);
    }
}

File: src/main/java/org/example/PathFinder.py
def solution(A):
    m = len(A)
    n = len(A[0])

    # initializing a
    B = [[
        0
        for _ in range(n)
    ]   for _ in range(m)
    ]

    for i in range(m):
        for j in range(n):

            if A[i][j] != '.': B[i][j] = 1

            if A[i][j] == 'A':
                B[i][j] = 0
                starti, startj = i,j

            if A[i][j] == 'v':
                k = i+1
                while k < m and A[k][j] == '.':
                    B[k][j] = 1
                    k += 1

            if A[i][j] == '^':
                k = i-1
                while k >= 0 and A[k][j] == '.':
                    B[k][j] = 1
                    k -= 1

            if A[i][j] == '>':
                k = j+1
                while k < n and A[i][k] == '.':
                    B[i][k] = 1
                    k += 1

            if A[i][j] == '<':
                k = j-1
                while k >= 0 and A[i][k] == '.':
                    B[i][k] = 1
                    k -= 1


    def f(i,j, visited):

        if i < 0: return False
        if j < 0: return False
        if i >= m: return False
        if j >= n: return False
        if (i,j) in visited: return False
        if B[i][j] == 1    : return False
        if j       == n-1  : return True
        visited.add((i,j))

        return (
            f(i,j+1, visited)
         or f(i,j-1, visited)
         or f(i+1,j, visited)
         or f(i-1,j, visited)
        )

    return f(starti, startj, set())